File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/components/utility/ImageSource.kt
package com.crescentflare.piratesgame.components.utility

import android.content.Context
import android.graphics.drawable.Animatable
import android.graphics.drawable.Drawable
import android.support.v4.content.ContextCompat
import android.widget.ProgressBar
import com.crescentflare.piratesgame.infrastructure.appconfig.CustomAppConfigManager
import com.crescentflare.piratesgame.infrastructure.coreextensions.urlDecode
import com.crescentflare.piratesgame.infrastructure.coreextensions.urlEncode
import com.crescentflare.piratesgame.infrastructure.imagegenerators.FilledOvalGenerator
import com.crescentflare.piratesgame.infrastructure.imagegenerators.FilledRectGenerator
import com.crescentflare.piratesgame.infrastructure.imagegenerators.StrokedPathGenerator
import com.crescentflare.piratesgame.infrastructure.inflator.Inflators


/**
 * Component utility: defines the source of an image (an internal or external image)
 */
class ImageSource {

    // --
    // Statics
    // --

    companion object {

        // --
        // Static: factory method
        // --

        fun fromObject(value: Any?): ImageSource? {
            if (value is String) {
                return ImageSource(value)
            } else if (value is Map<*, *>) {
                val result: Map<String, Any>? = Inflators.viewlet.mapUtil.asStringObjectMap(value)
                if (result != null) {
                    return ImageSource(result)
                }
            }
            return null
        }


        // --
        // Static: cache
        // --

        private val cachedDrawables = mutableMapOf<String, Drawable>()

    }


    // --
    // Members
    // --

    var type = Type.Unknown
    var parameters = mutableMapOf<String, Any>()
    var pathComponents = emptyList<String>()
    var otherSources = mutableListOf<ImageSource>()


    // --
    // Initialization
    // --

    constructor(value: Any?) {
        if (value is String) {
            initParse(value)
        } else if (value is Map<*, *>) {
            val result: Map<String, Any>? = Inflators.viewlet.mapUtil.asStringObjectMap(value)
            initParse(result ?: mapOf())
        }
    }

    constructor(string: String) {
        initParse(string)
    }

    constructor(map: Map<String, Any>) {
        initParse(map)
    }

    private fun initParse(string: String) {
        // Extract scheme
        var checkString = string
        val schemeMarker = checkString.indexOf("://")
        if (schemeMarker >= 0) {
            type = Type.fromString(checkString.substring(0, schemeMarker))
            checkString = checkString.substring(schemeMarker + 3)
        }

        // Extract parameters
        val paramMarker = checkString.indexOf('?')
        if (paramMarker >= 0) {
            // Get parameter string
            val paramString = checkString.substring(paramMarker + 1)
            checkString = checkString.substring(0, paramMarker)

            // Split into separate parameters and fill dictionary
            val paramItems = paramString.split("&")
            for (paramItem in paramItems) {
                val paramSet = paramItem.split("=")
                if (paramSet.size == 2) {
                    val key = paramSet[0].urlDecode()
                    if (key == "otherSources") {
                        val otherSourcesStringArray = paramSet[1].split(",")
                        for (otherSourcesString in otherSourcesStringArray) {
                            val otherSource = ImageSource(otherSourcesString.urlDecode())
                            otherSources.add(otherSource)
                        }
                    } else {
                        parameters[key] = paramSet[1].urlDecode()
                    }
                }
            }
        }

        // Finally set path to the remaining string
        pathComponents = checkString.split("/")
    }

    private fun initParse(map: Map<String, Any>) {
        type = Type.fromString(map["type"] as? String)
        val path = map["path"] ?: map["name"]
        if (path is String) {
            pathComponents = path.split("/")
        }
        val otherSources = Inflators.viewlet.mapUtil.optionalObjectList(map, "otherSources")
        for (otherSource in otherSources) {
            val imageSource = ImageSource.fromObject(otherSource)
            if (imageSource != null) {
                this.otherSources.add(imageSource)
            }
        }
        for (key in map.keys) {
            if (key != "type" && key != "path" && key != "name" && key != "otherSources") {
                val value = map[key]
                if (value != null) {
                    parameters[key] = value
                }
            }
        }
    }


    // --
    // Extract values
    // --

    val fullPath: String
        get() = pathComponents.joinToString("/")

    val onlineUri: String?
        get() {
            if (type == Type.OnlineImage || type == Type.SecureOnlineImage || type == Type.DevServerImage) {
                var uri = "${type.value}://$fullPath"
                if (type == Type.DevServerImage) {
                    uri = CustomAppConfigManager.currentConfig().devServerUrl
                    if (!uri.startsWith("http")) {
                        uri = "http://$uri"
                    }
                    uri = "$uri/pageimages/$fullPath"
                }
                getParameterString(listOf("caching", "colorize", "threePatch", "ninePatch"), true)?.let {
                    uri += "?$it"
                }
                return uri
            }
            return null
        }

    val tintColor: Int
        get() = Inflators.viewlet.mapUtil.optionalColor(parameters.toMap(), "colorize", 0)

    val threePatch: Int
        get() = Inflators.viewlet.mapUtil.optionalDimension(parameters.toMap(), "threePatch", -1)

    val ninePatch: Int
        get() = Inflators.viewlet.mapUtil.optionalDimension(parameters.toMap(), "ninePatch", -1)


    // --
    // Caching
    // --

    val caching: Caching
        get() = Caching.fromString(Inflators.viewlet.mapUtil.optionalString(parameters.toMap(), "caching", ""))

    val cacheKey: String
        get() {
            var uri = "${type.value}://$fullPath"
            getParameterString(listOf("caching", "colorize", "threePatch", "ninePatch"))?.let {
                uri += "?$it"
            }
            return uri
        }


    // --
    // Conversion
    // --

    val uri: String
        get() {
            var uri = "${type.value}://$fullPath"
            getParameterString()?.let {
                uri += "?$it"
            }
            return uri
        }

    val map: Map<String, Any>
        get() {
            val map = mutableMapOf<String, Any>(Pair("type", type.value), Pair("path", fullPath))
            map.putAll(parameters)
            if (otherSources.isNotEmpty()) {
                map["otherSources"] = otherSources.map { it.map }
            }
            return map
        }


    // --
    // Helper
    // --

    fun getDrawable(context: Context): Drawable? {
        var result: Drawable? = null
        if (caching == Caching.Always) {
            val cachedDrawable = cachedDrawables.get(cacheKey)
            if (cachedDrawable != null) {
                return cachedDrawable
            }
        }
        if (type == Type.InternalImage) {
            val resourceId = context.resources.getIdentifier(fullPath, "drawable", context.packageName)
            if (resourceId > 0) {
                result = ContextCompat.getDrawable(context, resourceId)
            }
        } else if (type == Type.SystemImage) {
            return SystemImageSource.load(context, fullPath)
        } else if (type == Type.Generate) {
            result = getGeneratedDrawable(context)
        }
        for (otherSource in otherSources) {
            val generatedImage = otherSource.getGeneratedDrawable(context, result)
            if (generatedImage != null) {
                result = generatedImage
            }
        }
        if (caching == Caching.Always && result != null) {
            cachedDrawables.put(cacheKey, result)
        }
        return result
    }

    private fun getGeneratedDrawable(context: Context, onDrawable: Drawable? = null): Drawable? {
        val generateType = GenerateType.fromString(fullPath)
        if (generateType != GenerateType.Unknown) {
            return when (generateType) {
                GenerateType.FilledRect -> FilledRectGenerator().generate(context, parameters, onDrawable)
                GenerateType.FilledOval -> FilledOvalGenerator().generate(context, parameters, onDrawable)
                GenerateType.StrokedPath -> StrokedPathGenerator().generate(context, parameters, onDrawable)
                else -> null
            }
        }
        return null
    }

    private fun getParameterString(ignoreParams: List<String> = emptyList(), ignoreOtherSources: Boolean = false): String? {
        if (parameters.isNotEmpty()) {
            var parameterString = ""
            for (key in parameters.keys.sorted()) {
                var ignore = false
                for (ignoreParam in ignoreParams) {
                    if (key == ignoreParam) {
                        ignore = true
                        break
                    }
                }
                val stringValue = Inflators.viewlet.mapUtil.optionalString(parameters, key, null)
                if (!ignore && stringValue != null) {
                    if (parameterString.isNotEmpty()) {
                        parameterString += "&"
                    }
                    parameterString += key.urlEncode() + "=" + stringValue.urlEncode()
                }
            }
            if (otherSources.isNotEmpty() && !ignoreOtherSources) {
                var otherSourceString = ""
                for (otherSource in otherSources) {
                    val otherSourceURI = otherSource.uri
                    if (otherSourceString.isNotEmpty()) {
                        otherSourceString += ","
                    }
                    otherSourceString += otherSourceURI.urlEncode()
                }
                if (parameterString.isNotEmpty()) {
                    parameterString += "&"
                }
                parameterString += "otherSources=$otherSourceString"
            }
            if (parameterString.length > 0) {
                return parameterString
            }
        }
        return null
    }


    // --
    // Type enums
    // --

    enum class Type(val value: String) {

        Unknown("unknown"),
        OnlineImage("http"),
        SecureOnlineImage("https"),
        DevServerImage("devserver"),
        InternalImage("app"),
        SystemImage("system"),
        Generate("generate");

        companion object {

            fun fromString(string: String?): Type {
                for (enum in Type.values()) {
                    if (enum.value == string) {
                        return enum
                    }
                }
                return Unknown
            }

        }

    }

    enum class GenerateType(val value: String) {

        Unknown("unknown"),
        FilledRect("filledRect"),
        FilledOval("filledOval"),
        StrokedPath("strokedPath");

        companion object {

            fun fromString(string: String?): GenerateType {
                for (enum in GenerateType.values()) {
                    if (enum.value == string) {
                        return enum
                    }
                }
                return Unknown
            }

        }

    }


    // --
    // Caching enum
    // --

    enum class Caching(val value: String) {

        Unknown("unknown"),
        Always("always"),
        Never("never");

        companion object {

            fun fromString(string: String?): Caching {
                for (enum in Caching.values()) {
                    if (enum.value == string) {
                        return enum
                    }
                }
                return Unknown
            }

        }

    }

}

File: pirates-ios/PiratesGame/Components/Utility/ImageSource.swift
//
//  ImageSource.swift
//  Component utility: defines the source of an image (an internal or external image)
//

import UIKit
import UniLayout

enum ImageSourceType: String {
    
    case unknown = "unknown"
    case onlineImage = "http"
    case secureOnlineImage = "https"
    case devServerImage = "devserver"
    case internalImage = "app"
    case systemImage = "system"
    case generate = "generate"
    
}

enum ImageSourceCaching: String {
    
    case unknown = "unknown"
    case always = "always"
    case never = "never"
    
}

enum ImageSourceGenerateType: String {
    
    case unknown = "unknown"
    case filledRect = "filledRect"
    case filledOval = "filledOval"
    case strokedPath = "strokedPath"
    
}

class ImageSource {
    
    // --
    // MARK: Members
    // --

    var type = ImageSourceType.unknown
    var parameters = [String: Any]()
    var pathComponents = [String]()
    var otherSources = [ImageSource]()
    

    // --
    // MARK: Initialization
    // --
    
    convenience init?(value: Any?) {
        if let stringValue = value as? String {
            self.init(string: stringValue)
        } else if let dictValue = value as? [String: Any] {
            self.init(dict: dictValue)
        } else {
            return nil
        }
    }

    init?(string: String?) {
        if let string = string {
            // Extract scheme
            var checkString = string
            if let schemeMarker = checkString.range(of: "://") {
                type = ImageSourceType(rawValue: String(checkString[..<schemeMarker.lowerBound])) ?? ImageSourceType.unknown
                checkString = String(checkString[schemeMarker.upperBound...])
            }
            
            // Extract parameters
            if let paramMarker = checkString.range(of: "?") {
                // Get parameter string
                let paramString = String(checkString[paramMarker.upperBound...])
                checkString = String(checkString[..<paramMarker.lowerBound])
                
                // Split into separate parameters and fill dictionary
                let paramItems = paramString.split(separator: "&").map(String.init)
                for paramItem in paramItems {
                    let paramSet = paramItem.split(separator: "=").map(String.init)
                    if paramSet.count == 2 {
                        let key = paramSet[0].urlDecode()
                        if key == "otherSources" {
                            let otherSourcesStringArray = paramSet[1].split(separator: ",").map(String.init)
                            for otherSourcesString in otherSourcesStringArray {
                                if let otherSource = ImageSource(string: otherSourcesString.urlDecode()) {
                                    otherSources.append(otherSource)
                                }
                            }
                        } else {
                            parameters[key] = paramSet[1].urlDecode()
                        }
                    }
                }
            }
            
            // Finally set path to the remaining string
            pathComponents = checkString.split(separator: "/").map(String.init)
        } else {
            return nil
        }
    }
    
    init(dict: [String: Any]) {
        let convUtil = Inflators.viewlet.convUtil
        type = ImageSourceType(rawValue: convUtil.asString(value: dict["type"]) ?? "unknown") ?? .unknown
        if let path = convUtil.asString(value: dict["path"]) ?? convUtil.asString(value: dict["name"]) {
            pathComponents = path.split(separator: "/").map(String.init)
        }
        if let otherSources = dict["otherSources"] as? [Any] {
            for otherSource in otherSources {
                if let source = ImageSource(value: otherSource) {
                    self.otherSources.append(source)
                }
            }
        }
        for (key, value) in dict {
            if key != "type" && key != "path" && key != "name" && key != "otherSources" {
                parameters[key] = value
            }
        }
    }

    
    // --
    // MARK: Extract values
    // --
    
    var fullPath: String {
        get {
            return pathComponents.joined(separator: "/")
        }
    }
    
    var onlineUri: String? {
        get {
            if type == .onlineImage || type == .secureOnlineImage || type == .devServerImage {
                var uri = "\(type.rawValue)://\(fullPath)"
                if type == .devServerImage {
                    uri = CustomAppConfigManager.currentConfig().devServerUrl
                    if !uri.hasPrefix("http") {
                        uri = "http://\(uri)"
                    }
                    uri = "\(uri)/pageimages/\(fullPath)"
                }
                if let paramString = getParameterString(ignoreParams: ["caching", "colorize", "threePatch", "ninePatch"], ignoreOtherSources: true) {
                    uri += "?" + paramString
                }
                return uri
            }
            return nil
        }
    }

    var tintColor: UIColor? {
        get {
            if let colorizeString = parameters["colorize"] {
                return Inflators.viewlet.convUtil.asColor(value: colorizeString)
            }
            return nil
        }
    }
    
    var threePatch: CGFloat? {
        get {
            return Inflators.viewlet.convUtil.asDimension(value: parameters["threePatch"])
        }
    }
    
    var ninePatch: CGFloat? {
        get {
            return Inflators.viewlet.convUtil.asDimension(value: parameters["ninePatch"])
        }
    }
    

    // --
    // MARK: Caching
    // --
    
    private static var cachedImages = [String: UIImage]()

    var caching: ImageSourceCaching {
        get {
            return ImageSourceCaching(rawValue: Inflators.viewlet.convUtil.asString(value: parameters["caching"]) ?? "") ?? .unknown
        }
    }
    
    var cacheKey: String {
        get {
            var uri = "\(type.rawValue)://\(fullPath)"
            if let paramString = getParameterString(ignoreParams: ["caching", "colorize", "threePatch", "ninePatch"]) {
                uri += "?" + paramString
            }
            return uri
        }
    }
    

    // --
    // MARK: Conversion
    // --

    var uri: String {
        get {
            var uri = "\(type.rawValue)://\(fullPath)"
            if let paramString = getParameterString() {
                uri += "?" + paramString
            }
            return uri
        }
    }
    
    var dictionary: [String: Any] {
        get {
            var dictionary: [String: Any] = ["type": type.rawValue, "path": fullPath]
            for (key, value) in parameters {
                dictionary[key] = value
            }
            if otherSources.count > 0 {
                dictionary["otherSources"] = otherSources.map { $0.dictionary }
            }
            return dictionary
        }
    }
    
    
    // --
    // MARK: Helper
    // --

    func getImage() -> UIImage? {
        var result: UIImage?
        if caching == .always, let cachedImage = ImageSource.cachedImages[cacheKey] {
            return cachedImage
        }
        if type == .internalImage {
            let bundle = Bundle(for: ImageSource.self)
            let path = fullPath
            if path.hasPrefix("unilayout-") {
                result = UniAssets.obtainAsset(name: path.replacingOccurrences(of: "unilayout-", with: ""))
            } else {
                result = UIImage(named: path, in: bundle, compatibleWith: nil)
            }
        } else if type == .systemImage {
            result = SystemImageSource.load(path: fullPath)
        } else if type == .generate {
            result = getGeneratedImage()
        }
        for otherSource in otherSources {
            if let generatedImage = otherSource.getGeneratedImage(onImage: result) {
                result = generatedImage
            }
        }
        if caching == .always, let imageResult = result {
            ImageSource.cachedImages[cacheKey] = imageResult
        }
        return result
    }
    
    private func getGeneratedImage(onImage: UIImage? = nil) -> UIImage? {
        if let generateType = ImageSourceGenerateType(rawValue: fullPath) {
            switch generateType {
            case .filledRect:
                return FilledRectGenerator().generate(attributes: parameters, onImage: onImage)
            case .filledOval:
                return FilledOvalGenerator().generate(attributes: parameters, onImage: onImage)
            case .strokedPath:
                return StrokedPathGenerator().generate(attributes: parameters, onImage: onImage)
            default:
                break
            }
        }
        return nil
    }
    
    private func getParameterString(ignoreParams: [String] = [], ignoreOtherSources: Bool = false) -> String? {
        if parameters.count > 0 {
            var parameterString = ""
            for key in parameters.keys.sorted() {
                var ignore = false
                for ignoreParam in ignoreParams {
                    if key == ignoreParam {
                        ignore = true
                        break
                    }
                }
                if !ignore, let value = Inflators.viewlet.convUtil.asString(value: parameters[key]) {
                    if parameterString.count > 0 {
                        parameterString += "&"
                    }
                    parameterString += key.urlEncode() + "=" + value.urlEncode()
                }
            }
            if otherSources.count > 0 && !ignoreOtherSources {
                var otherSourceString = ""
                for otherSource in otherSources {
                    let otherSourceURI = otherSource.uri
                    if otherSourceString.count > 0 {
                        otherSourceString += ","
                    }
                    otherSourceString += otherSourceURI.urlEncode()
                }
                if parameterString.count > 0 {
                    parameterString += "&"
                }
                parameterString += "otherSources=" + otherSourceString
            }
            if parameterString.count > 0 {
                return parameterString
            }
        }
        return nil
    }
    
}
